### ISSUE TITLE
Discuss / Decide a nice API to allow Context to do things on instance instantiation

### ISSUE THREAD

**Issue by syrusakbary** [wasmerio/wasmer#2924]
We can have the context or the environment to have a method `on_new_instance` that will receive an instance and will modify the environment.

Kind of:
```rust
impl WASIContext {
    fn on_new_instance(&mut self, &instance) {
        let memory = instance.exports.get_memory("memory")?;
        self.data_mut().memory = Some(memory.clone());

        let start = instance.exports.get_function("_start")?;
        self.start_function = start;
    }
}

impl WASIContext {
    fn start(&self) -> Result<i32, Error> {
        self.start_function.call(&[])?;
    }
}
```
And then:

```rust
        let mut ctx = WASIContext::new(&store, env.clone());
        let imports = self.get_imports(&mut ctx.as_context_mut(), &module)?;
        let instance = Instance::new(&mut ctx, &module, &imports)?;

        ctx.start()
```

**Comment by syrusakbary** [comment]
We should have an architectural docs to discuss things before implementing

**Comment by syrusakbary** [comment]
Proposal: have a new function that takes an optional 

```rust
let env = FunctionEnv::new_with_init(&mut store, (), |store: StoreRef, instance: Instance| {
    // ...
});
```

**Comment by syrusakbary** [comment]
After working a bit on this, I realized that it's actually not trivial to implement. And to be honest after some research, I'm not sure if it's worth it as we have designed it atm.

Basically, implementing a solution would imply that the `wasmer::Function`  starts keeping track of the Function Type `<T>` (`T` being the environment).
However, all exportable elements must be static (that means, no generics). So that breaks the requirement of keeping the environment type somewhere.

There are always hacks around this, but after some extra thought I don't think they impose a good API to the user. Why?
Functions and Function Environments can be cloned and repeated, it's not obvious (as a user) when the init_with_env  function is called, what is supposed to do. For example, in WASI, it's only useful when binding "for the first time", as afterwards the `Option` will be non-empty and then non-replaced.

It adds unnecessary complexity with an API that I don't think is future proof-safe
I think we should keep thinking about ways to approach it. But it needs to be both simple to write and simple to understand
